lib/services/role.go

```diff
@@ -118,41 +132,330 @@ type Access interface {
+	r.Spec.Options[optionName] = optionValue
+}
+
+// GetLogins gets system logins for allow or deny condition.
+func (r *RoleV3) GetLogins(rct RoleConditionType) []string {
+	if rct == Allow {
+		return r.Spec.Allow.Logins
+	}
+	return r.Spec.Deny.Logins
+}
+
+// SetLogins sets system logins for allow or deny condition.
+func (r *RoleV3) SetLogins(rct RoleConditionType, logins []string) {
+	if rct == Allow {
+		r.Spec.Allow.Logins = logins
+	} else {
+		r.Spec.Deny.Logins = logins
+	}
+}
+
+// GetNamespaces gets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) GetNamespaces(rct RoleConditionType) []string {
+	if rct == Allow {
+		return r.Spec.Allow.Namespaces
+	}
+	return r.Spec.Deny.Namespaces
+}
+
+// GetNamespaces sets a list of namespaces this role is allowed or denied access to.
+func (r *RoleV3) SetNamespaces(rct RoleConditionType, namespaces []string) {
+	if rct == Allow {
+		r.Spec.Allow.Namespaces = namespaces
+	} else {
+		r.Spec.Deny.Namespaces = namespaces
+	}
+}
+
+// GetNodeLabels gets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) GetNodeLabels(rct RoleConditionType) map[string]string {
+	if rct == Allow {
+		return r.Spec.Allow.NodeLabels
+	}
+	return r.Spec.Deny.NodeLabels
+}
+
+// SetNodeLabels sets the map of node labels this role is allowed or denied access to.
+func (r *RoleV3) SetNodeLabels(rct RoleConditionType, labels map[string]string) {
+	if rct == Allow {
+		r.Spec.Allow.NodeLabels = labels
+	} else {
+		r.Spec.Deny.NodeLabels = labels
+	}
+}
+
+// GetRules gets all allow or deny rules.
+func (r *RoleV3) GetRules(rct RoleConditionType) []RoleRule {
+	if rct == Allow {
+		return r.Spec.Allow.Rules
+	}
+	return r.Spec.Deny.Rules
+}
+
+// SetRules sets an allow or deny rule.
+func (r *RoleV3) SetRule(rct RoleConditionType, resources []string, verbs []string) {
+	rr := RoleRule{
+		Resources: resources,
+		Verbs:     verbs,
+	}
+	if rct == Allow {
+		r.Spec.Allow.Rules = append(r.Spec.Allow.Rules, rr)
+	} else {
+		r.Spec.Deny.Rules = append(r.Spec.Deny.Rules, rr)
+	}
+}
+
+// Check checks validity of all parameters and sets defaults
+func (r *RoleV3) CheckAndSetDefaults() error {
+	// make sure we have defaults for all fields
+	if r.Metadata.Name == "" {
+		return trace.BadParameter("missing parameter Name")
+	}
+	if r.Metadata.Namespace == "" {
+		r.Metadata.Namespace = defaults.Namespace
+	}
+	if r.Spec.MaxSessionTTL.Duration == 0 {
+		r.Spec.MaxSessionTTL.Duration = defaults.MaxCertDuration
+	}
+	if r.Spec.MaxSessionTTL.Duration < defaults.MinCertDuration {
+		return trace.BadParameter("maximum session TTL can not be less than")
+	}
+	if r.Spec.Allow.Namespaces == nil {
+		r.Spec.Allow.Namespaces = []string{defaults.Namespace}
+	}
+	if r.Spec.Allow.NodeLabels == nil {
+		r.Spec.Allow.NodeLabels = map[string]string{Wildcard: Wildcard}
+	}
+	if r.Spec.Allow.Rules == nil {
+		r.Spec.Allow.Rules = []RoleRule{
+			RoleRule{
+				Resources: []string{
```

Comment:
This default set of rules: what is it? I.e. how would you describe it? It's some kind of implicit rule set we need to name and declare (and assign by making a copy).